service/kernel/oneapi/cl_kernels/partition.cl

```diff
@@ -0,0 +1,198 @@
+
+        int iStart = group_id * nElementsForSubgroup;
+        int iEnd   = (group_id + 1) * nElementsForSubgroup;
+
+        if (iEnd > nElems)
+        {
+            iEnd = nElems;
+        }
+
+        int sum = 0;
+
+        for (int i = iStart + local_id; i < iEnd; i += local_size)
+        {
+            const int value = mask[data[i]];
+            sum += sub_group_reduce_add(value);
+        }
+
+        if (local_id == 0)
+        {
+            partialSums[group_id] = sum;
+        }
+    }
+
+    __kernel void sumScan(const __global int * partialSums, __global int * partialPrefixSums, __global int * totalSum, int nSubgroupSums) {
+        if (get_sub_group_id() > 0) return;
+
+        const int local_size = get_sub_group_size();
+        const int local_id   = get_sub_group_local_id();
+
+        int sum = 0;
+
+        for (int i = local_id; i < nSubgroupSums; i += local_size)
+        {
+            int value            = partialSums[i];
+            int boundary         = sub_group_scan_exclusive_add(value);
+            partialPrefixSums[i] = sum + boundary;
+            sum += sub_group_reduce_add(value);
+        }
+
+        if (local_id == 0)
+        {
+            totalSum[0]                      = sum;
+            partialPrefixSums[nSubgroupSums] = sum;
+        }
+    }
+
+    __kernel void reorder(const __global int * mask, const __global algorithmFPType * data, __global algorithmFPType * outData,
+                          const __global int * partialPrefixSums, int nElems) {
+        const int n_groups             = get_num_groups(0);
+        const int n_sub_groups         = get_num_sub_groups();
+        const int n_total_sub_groups   = n_sub_groups * n_groups;
+        const int nElementsForSubgroup = nElems / n_total_sub_groups + !!(nElems % n_total_sub_groups);
+        const int local_size           = get_sub_group_size();
+
+        const int id           = get_local_id(0);
+        const int local_id     = get_sub_group_local_id();
+        const int sub_group_id = get_sub_group_id();
+        const int group_id     = get_group_id(0) * n_sub_groups + sub_group_id;
+
+        int iStart = group_id * nElementsForSubgroup;
+        int iEnd   = (group_id + 1) * nElementsForSubgroup;
+
+        if (iEnd > nElems)
+        {
+            iEnd = nElems;
+        }
+
+        int groupOffset = partialPrefixSums[group_id];
+        int sum         = 0;
+
+        for (int i = iStart + local_id; i < iEnd; i += local_size)
+        {
+            int part     = mask[i];
+            int boundary = groupOffset + sum + sub_group_scan_exclusive_add(part);
+            if (part)
+            {
+                const int pos_new = boundary;
+                outData[pos_new]  = data[i];
+            }
+            sum += sub_group_reduce_add(part);
+        }
+    }
+
+    __kernel void reorderIndex(const __global int * mask, const __global int * data, __global int * outData, const __global int * partialPrefixSums,
```

Comment:
Same concerns as for previous kernel:
1. Use another word instead of "data" when you actually mean indices.
2. Try to avoid branching in main loop.